File: src/002.String/299.BullsAndCows/299.猜数字游戏.go
/*
 * @lc app=leetcode.cn id=299 lang=golang
 *
 * [299] 猜数字游戏
 */

// @lc code=start
func getHint(secret string, guess string) string {
	result := ""
	A, B := 0, 0
	secretArr := make([]int, 10)
	guessArr := make([]int, 10)

	for i, v := range secret {
		secretNum, _ := strconv.Atoi(string(v))
		guessNum, _ := strconv.Atoi(string(guess[i]))

		if secretNum == guessNum {
			A++
		} else {
			secretArr[secretNum]++
			guessArr[guessNum]++
		}
	}

	for i := 0; i < 10; i++ {
		B += min(secretArr[i], guessArr[i])
	}

	result += strconv.Itoa(A) + "A" + strconv.Itoa(B) + "B"

	return result
}

func min(a int, b int) int {
	if a < b {
		return a
	}
	return b
}

// @lc code=end

File: src/002.String/299.BullsAndCows/299.猜数字游戏.js
/*
 * @lc app=leetcode.cn id=299 lang=javascript
 *
 * [299] 猜数字游戏
 */

// @lc code=start
/**
 * @param {string} secret
 * @param {string} guess
 * @return {string}
 */
var getHint = function (secret, guess) {
	let A = 0,
		B = 0,
		secretObj = {},
		guessObj = {};

	for (let i = 0; i < secret.length; i++) {
		if (secret[i] === guess[i]) {
			A++;
		} else {
			secretObj[secret[i]] = (secretObj[secret[i]] || 0) + 1;
			guessObj[guess[i]] = (guessObj[guess[i]] || 0) + 1;
		}
	}

	let usedGuess = [];
	for (let i = 0; i < guess.length; i++) {
		if (!usedGuess.includes(guess[i])) {
			B += Math.min(secretObj[guess[i]] || 0, guessObj[guess[i]] || 0);
			usedGuess.push(guess[i]);
		}
	}

	return `${A}A${B}B`;
};
// console.log(getHint('1122', '2211'));
// @lc code=end
